roofit/roofitcore/test/testRunContextTracker.cxx

```diff
@@ -0,0 +1,215 @@
+   RunContextTracker runContextTracker{runContext};
+   {
+      runContextTracker.resetTrackers();
+
+      changeVal(a0);
+
+      // The first argument is just to have a caller for the RooMsgService
+      runContextTracker.cleanRunContext(model, runContext);
+
+      // Let's think about what the runContextTracker should have cleared
+      // based on the computation graph topology
+      ASSERT_FALSE(hasKept(runContext, a0));
+      ASSERT_TRUE(hasKept(runContext, a1));
+      ASSERT_FALSE(hasKept(runContext, fy));
+      ASSERT_TRUE(hasKept(runContext, sigma));
+      ASSERT_FALSE(hasKept(runContext, model));
+      ASSERT_TRUE(hasKept(runContext, x));
+      ASSERT_TRUE(hasKept(runContext, y));
+   }
+   {
+      // 2nd iteration
+      model.getValues(runContext, &nset);
+      runContextTracker.resetTrackers();
+
+      changeVal(a0);
+      changeVal(a1);
+      changeVal(sigma);
+
+      runContextTracker.cleanRunContext(model, runContext);
+
+      ASSERT_FALSE(hasKept(runContext, a0));
+      ASSERT_FALSE(hasKept(runContext, a1));
+      ASSERT_FALSE(hasKept(runContext, fy));
+      ASSERT_FALSE(hasKept(runContext, sigma));
+      ASSERT_FALSE(hasKept(runContext, model));
+      ASSERT_TRUE(hasKept(runContext, x));
+      ASSERT_TRUE(hasKept(runContext, y));
+   }
+   {
+      // 3rd iteration
+      model.getValues(runContext, &nset);
+      runContextTracker.resetTrackers();
+
+      changeVal(sigma);
+
+      runContextTracker.cleanRunContext(model, runContext);
+
+      ASSERT_TRUE(hasKept(runContext, a0));
+      ASSERT_TRUE(hasKept(runContext, a1));
+      ASSERT_TRUE(hasKept(runContext, fy));
+      ASSERT_FALSE(hasKept(runContext, sigma));
+      ASSERT_FALSE(hasKept(runContext, model));
+      ASSERT_TRUE(hasKept(runContext, x));
+      ASSERT_TRUE(hasKept(runContext, y));
+   }
+}
+
+TEST(RooRunContextTracker, InsideRooNLLVar)
+{
+   // It is important to also test the RooRunContextTracker as it is used
+   // inside RooNLLVar, because there are many things that can go wrong when
+   // the operation modes are changed during the construction of the RooNLLVar
+
+   // Create observables
+   RooRealVar x("x", "x", -10, 10);
+   RooRealVar y("y", "y", -10, 10);
+
+   // Create function f(y) = a0 + a1*y
+   RooRealVar a0("a0", "a0", -0.5, -5, 5);
+   RooRealVar a1("a1", "a1", -0.5, -1, 1);
+   RooPolyVar fy("fy", "fy", y, RooArgSet(a0, a1));
+
+   // Create gauss(x,f(y),s)
+   RooRealVar sigma("sigma", "width of gaussian", 0.5, 0.1, 2.0);
+   RooGaussian model("model", "Gaussian with shifting mean", x, fy, sigma);
+
+   auto dataSet = makeFakeDataXY();
+
+   auto changeVal = [](RooRealVar &var) { var.setVal(var.getVal() + 0.001); };
+   auto hasKept = [](std::string const &logStr, RooAbsReal const &arg) {
+      return logStr.find(std::string(arg.GetName()) + " kept", 0) != std::string::npos;
+   };
+
+   // Create the NLL object
+   auto *nll = model.createNLL(dataSet, RooFit::BatchMode(true));
```

Comment:
Try to keep tests leak free, so one could run valgrind or address sanitizer on them if desired.
```suggestion
   std::unique_ptr<RooNLLVar> nll( model.createNLL(dataSet, RooFit::BatchMode(true)) );
```

Reply:
Done!